### PR TITLE
Allow FileMutex to be explicitly told it is operating on a Windows filesystem

### PR DESCRIPTION
**Pull request by brandonkelly**
| Q             | A
| ------------- | ---
| Is bugfix?    | yes
| New feature?  | no
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #16603

### CODE REVIEW

**Comment by rob006** [comment]
There is 6 other places where `DIRECTORY_SEPARATOR === '\\'` is used for Windows detection. Shouldn't we have some generic solution for this instead for patching this in one place?

**Comment by brandonkelly** [comment]
Only IO operations are going to be potentially affected by #16603 (not BaseConsole::streamSupportsAnsiColors(), for example).

Even if there are similar issues elsewhere, I think it’s important that FileMutex has a localized `isWindows` setting, because its base path is configurable. You could have a Yii application where only the base mutex directory is on a Windows share, but the rest of the app is not.

So this seems like a good starting point, and if another issue arrises later, we could look into an application config setting that determines how IO operations should be handled. All that would change in FileMutex at that point is how the default `isWindows` value is determined:

```php
// old:
$this->isWindows = DIRECTORY_SEPARATOR === '\\';

// new:
$this->isWindows = Yii::$app->ioIsWindows;
```

(or something like that)

File: framework/mutex/FileMutex.php [hunk 1]
```diff
@@ -59,6 +59,12 @@ class FileMutex extends Mutex
      * but read-only for other users.
      */
     public $dirMode = 0775;
+    /**
+     * @var bool whether file handling should assume a Windows file system.
+     * This value will determine how [[releaseLock()]] goes about deleting the lock file.
+     * If not set, it will be determined by checking the DIRECTORY_SEPARATOR constant.
```
Comment (samdark) [inline framework/mutex/FileMutex.php]: Need `@since`.

Comment (brandonkelly) [reply framework/mutex/FileMutex.php]: @samdark done

**Comment by brandonkelly** [comment]
Awesome, thanks!

**Pull request merged**